components/suggest/src/db.rs

```diff
@@ -1339,6 +1366,53 @@ impl<'a> SuggestDao<'a> {
     }
 }
 
+/// Helper struct to get full_keyword_ids for a suggestion
+///
+/// `FullKeywordInserter` handles repeated full keywords efficiently.  The first instance will
+/// cause a row to be inserted into the database.  Subsequent instances will return the same
+/// full_keyword_id.
+struct FullKeywordInserter<'a> {
+    conn: &'a Connection,
+    suggestion_id: i64,
+    last_inserted: Option<(&'a str, i64)>,
+}
+
+impl<'a> FullKeywordInserter<'a> {
+    fn new(conn: &'a Connection, suggestion_id: i64) -> Self {
+        Self {
+            conn,
+            suggestion_id,
+            last_inserted: None,
+        }
+    }
+
+    fn get_id(&mut self, full_keyword: &'a str) -> rusqlite::Result<i64> {
```

Comment:
Suggestion: how about renaming this to `maybe_insert()`? `get_id()` kind of hides the insert part.